### PR TITLE
Change analyze() to return Analysis

### PR DESCRIPTION
**Pull request by justinfagnani**
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: src/model/import.ts [hunk 1]
```diff
@@ -61,8 +61,8 @@ export class ScannedImport implements Resolvable {
     if (!document.analyzer.canResolveUrl(this.url)) {
       return;
     }
-    const importedDocument = document.analyzer._getDocument(this.url);
-    if (!importedDocument) {
+    const importedDocumentOrWarning = document.analyzer.getDocument(this.url);
```
Comment (justinfagnani) [inline src/model/import.ts]: rename document.analyzer to analysisContext

File: src/core/analysis-context.ts [hunk 2]
```diff
@@ -60,7 +65,7 @@ import {LanguageAnalyzer} from './language-analyzer';
  * appear to be statefull with respect to file updates.
  */
 export class AnalysisContext {
-  private _parsers = new Map<string, Parser<ParsedDocument<any, any>>>([
+  _parsers = new Map<string, Parser<ParsedDocument<any, any>>>([
```
Comment (rictic) [inline src/core/analysis-context.ts]: made public but readonly

File: src/core/analysis-context.ts [hunk 3]
```diff
@@ -77,13 +82,16 @@ export class AnalysisContext {
...
-  private _loader: UrlLoader;
-  private _resolver: UrlResolver;
+  _loader: UrlLoader;
+  _resolver: UrlResolver;
 
   private _cache = new AnalysisCache();
 
   private _generation = 0;
 
+  /** Resolves when analysis is complete */
```
Comment (rictic) [inline src/core/analysis-context.ts]: done

File: src/core/analysis-context.ts [hunk 4]
```diff
@@ -133,62 +141,105 @@ export class AnalysisContext {
...
-          end: {line: 0, column: 0}
-        },
-        code: 'unable-to-analyze',
-        message: `Unable to analyze file: ${e && e.message || e}`,
-        severity: Severity.ERROR
-      };
+  async analyze(urls: string[]): Promise<AnalysisContext> {
+    const resolvedUrls = urls.map((url) => this.resolveUrl(url));
+
+    // 1. Await current analysis if there is one, so we can check to see if has
```
Comment (rictic) [inline src/core/analysis-context.ts]: documented above

File: src/core/analysis-context.ts [hunk 5]
```diff
@@ -133,62 +141,105 @@ export class AnalysisContext {
...
-      } else {
-        warnings.push(docOrWarning);
-      }
-    }
-    return new Package(documents, warnings);
+  /**
+   * Internal analysis method called when we know we need to fork.
+   */
+  private async _analyze(resolvedUrls: string[]): Promise<AnalysisContext> {
+    // console.log('_analyze', resolvedUrls);
```
Comment (rictic) [inline src/core/analysis-context.ts]: done

File: src/core/analysis-context.ts [hunk 6]
```diff
@@ -197,15 +248,24 @@ export class AnalysisContext {
...
+  getDocument(url: string): Document | Warning {
     const resolvedUrl = this.resolveUrl(url);
     const cachedResult = this._cache.analyzedDocuments.get(resolvedUrl);
     if (cachedResult) {
       return cachedResult;
     }
     const scannedDocument = this._cache.scannedDocuments.get(resolvedUrl);
     if (!scannedDocument) {
-      return;
+      return <Warning>{
```
Comment (rictic) [inline src/core/analysis-context.ts]: Nah, we need a fallback anyways.

File: src/model/import.ts (hunk 1 above)
Comment (rictic) [inline src/model/import.ts]: done, also added an underscore to discourage touching.

**Comment by justinfagnani** [comment]
Please update CHANGELOG

**Comment by rictic** [comment]
Done

**Pull request merged**